File: r/tests/testthat/test-streampress.R
test_that("sparse round-trip preserves values", {
  library(Matrix)
  set.seed(42)
  m <- sparseMatrix(
    i = sample.int(100, 200, replace = TRUE),
    j = sample.int(50,  200, replace = TRUE),
    x = runif(200),
    dims = c(100, 50)
  )
  path <- tempfile(fileext = ".spz")
  st_write(m, path)
  m2 <- st_read(path)
  expect_equal(as.matrix(m), as.matrix(m2), tolerance = 1e-6)
})

test_that("column slice returns correct submatrix", {
  library(Matrix)
  set.seed(1)
  m <- sparseMatrix(
    i = sample.int(200, 500, replace = TRUE),
    j = sample.int(100, 500, replace = TRUE),
    x = runif(500),
    dims = c(200, 100)
  )
  path <- tempfile(fileext = ".spz")
  st_write(m, path, include_transpose = FALSE)

  sub <- st_slice_cols(path, cols = 11L:30L)
  ref <- m[, 11:30, drop = FALSE]
  expect_equal(as.matrix(ref), as.matrix(sub), tolerance = 1e-6)
  expect_equal(ncol(sub), 20L)
})

test_that("st_info returns correct dimensions", {
  library(Matrix)
  set.seed(7)
  m <- sparseMatrix(
    i = sample.int(50, 100, replace = TRUE),
    j = sample.int(20, 100, replace = TRUE),
    x = rpois(100, lambda = 3),
    dims = c(50, 20)
  )
  path <- tempfile(fileext = ".spz")
  st_write(m, path)
  info <- st_info(path)
  expect_equal(info$rows, 50L)
  expect_equal(info$cols, 20L)
})

test_that("obs and var metadata round-trips", {
  library(Matrix)
  set.seed(3)
  m <- sparseMatrix(
    i = sample.int(30, 60, replace = TRUE),
    j = sample.int(10, 60, replace = TRUE),
    x = runif(60),
    dims = c(30, 10)
  )
  obs <- data.frame(id = paste0("obs_", 1:30), stringsAsFactors = FALSE)
  var <- data.frame(id = paste0("var_", 1:10), stringsAsFactors = FALSE)
  path <- tempfile(fileext = ".spz")
  st_write(m, path, obs = obs, var = var)

  obs2 <- st_read_obs(path)
  var2 <- st_read_var(path)
  expect_equal(as.character(obs$id), as.character(obs2$id))
  expect_equal(as.character(var$id), as.character(var2$id))
})

test_that("dense round-trip preserves values", {
  set.seed(11)
  d <- matrix(rnorm(200 * 8), nrow = 200, ncol = 8)
  path <- tempfile(fileext = ".spz")
  st_write_dense(d, path)
  d2 <- st_read_dense(path)
  expect_equal(d, d2, tolerance = 1e-6)
})

test_that("integer matrix is handled correctly", {
  library(Matrix)
  set.seed(99)
  # Integer-valued sparse matrix (counts)
  m <- sparseMatrix(
    i = sample.int(40, 80, replace = TRUE),
    j = sample.int(15, 80, replace = TRUE),
    x = as.double(rpois(80, 2)),
    dims = c(40, 15)
  )
  path <- tempfile(fileext = ".spz")
  st_write(m, path)
  m2 <- st_read(path)
  expect_equal(as.matrix(m), as.matrix(m2), tolerance = 1e-9)
})

File: python/tests/test_streampress.py
"""
pytest tests for the streampress Python package.

All tests use synthetic data so they run anywhere with no external files.
"""

import numpy as np
import pytest
import scipy.sparse as sp
import tempfile
import os

import streampress as stp


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def make_sparse(m=500, n=100, density=0.05, seed=42):
    rng = np.random.default_rng(seed)
    return sp.random(m, n, density=density, format="csc",
                     dtype=np.float64, random_state=rng)


def make_count_matrix(m=200, n=50, seed=0):
    """Integer count matrix (like scRNA-seq UMI counts)."""
    rng = np.random.default_rng(seed)
    A = sp.random(m, n, density=0.1, format="csc",
                  random_state=rng)
    A.data = np.floor(rng.exponential(5.0, size=A.nnz)) + 1.0
    return A


# ---------------------------------------------------------------------------
# Round-trip tests
# ---------------------------------------------------------------------------

class TestRoundTrip:

    def test_basic_roundtrip(self, tmp_path):
        A = make_sparse()
        path = str(tmp_path / "test.spz")
        stp.st_write(A, path, include_transpose=False)
        B = stp.st_read(path)
        assert B.shape == A.shape
        assert B.nnz == A.nnz
        np.testing.assert_allclose(B.toarray(), A.toarray(), rtol=1e-5)

    def test_roundtrip_with_transpose(self, tmp_path):
        A = make_sparse(m=200, n=80)
        path = str(tmp_path / "test_t.spz")
        stp.st_write(A, path, include_transpose=True)
        B = stp.st_read(path)
        assert B.shape == A.shape
        np.testing.assert_allclose(B.toarray(), A.toarray(), rtol=1e-5)

    def test_count_matrix(self, tmp_path):
        A = make_count_matrix()
        path = str(tmp_path / "counts.spz")
        stp.st_write(A, path, precision="fp32", include_transpose=False)
        B = stp.st_read(path)
        assert B.shape == A.shape
        # fp32 is lossless for integers up to 2^24
        np.testing.assert_allclose(B.toarray(), A.toarray(), rtol=1e-4)

    def test_single_column(self, tmp_path):
        A = make_sparse(m=100, n=1)
        path = str(tmp_path / "single.spz")
        stp.st_write(A, path, include_transpose=False)
        B = stp.st_read(path)
        assert B.shape == A.shape

    def test_empty_matrix(self, tmp_path):
        A = sp.csc_matrix((50, 20))
        path = str(tmp_path / "empty.spz")
        stp.st_write(A, path, include_transpose=False)
        B = stp.st_read(path)
        assert B.shape == A.shape
        assert B.nnz == 0

    @pytest.mark.parametrize("precision", ["auto", "fp32", "fp64"])
    def test_precision_roundtrip(self, tmp_path, precision):
        A = make_sparse(m=100, n=30)
        path = str(tmp_path / f"prec_{precision}.spz")
        stp.st_write(A, path, precision=precision, include_transpose=False)
        B = stp.st_read(path)
        np.testing.assert_allclose(B.toarray(), A.toarray(), rtol=1e-4)


# ---------------------------------------------------------------------------
# Info tests
# ---------------------------------------------------------------------------

class TestInfo:

    def test_info_dimensions(self, tmp_path):
        A = make_sparse(m=300, n=50)
        path = str(tmp_path / "info.spz")
        stp.st_write(A, path, include_transpose=False)
        info = stp.st_info(path)
        assert info["nrow"] == 300
        assert info["ncol"] == 50
        assert info["nnz"] == A.nnz
        assert info["version"] >= 2
        assert not info["has_transpose"]

    def test_info_with_transpose(self, tmp_path):
        A = make_sparse(m=100, n=40)
        path = str(tmp_path / "info_t.spz")
        stp.st_write(A, path, include_transpose=True)
        info = stp.st_info(path)
        assert info["has_transpose"]


# ---------------------------------------------------------------------------
# Column slicing
# ---------------------------------------------------------------------------

class TestSlice:

    def test_slice_cols_contiguous(self, tmp_path):
        A = make_sparse(m=200, n=100)
        path = str(tmp_path / "slice.spz")
        stp.st_write(A, path, include_transpose=False)
        B = stp.st_slice_cols(path, list(range(10, 30)))
        assert B.shape == (200, 20)
        np.testing.assert_allclose(
            B.toarray(), A[:, 10:30].toarray(), rtol=1e-4)

    def test_slice_rows(self, tmp_path):
        A = make_sparse(m=200, n=50)
        path = str(tmp_path / "slice_r.spz")
        stp.st_write(A, path, include_transpose=True)
        B = stp.st_slice_rows(path, list(range(0, 50)))
        assert B.shape[0] == 50
        assert B.shape[1] == 50

    def test_slice_combined(self, tmp_path):
        A = make_sparse(m=200, n=100)
        path = str(tmp_path / "slice_comb.spz")
        stp.st_write(A, path, include_transpose=True)
        B = stp.st_slice(path, cols=list(range(20, 40)))
        assert B.shape[1] == 20


# ---------------------------------------------------------------------------
# Dense round-trip
# ---------------------------------------------------------------------------

class TestDense:

    def test_dense_roundtrip(self, tmp_path):
        rng = np.random.default_rng(42)
        X = rng.standard_normal((100, 50))
        path = str(tmp_path / "dense.spz")
        stp.st_write_dense(X, path, codec=0)
        Y = stp.st_read_dense(path)
        np.testing.assert_allclose(Y, X, rtol=1e-5)

    def test_dense_shape(self, tmp_path):
        X = np.ones((30, 20))
        path = str(tmp_path / "dense_ones.spz")
        stp.st_write_dense(X, path)
        Y = stp.st_read_dense(path)
        assert Y.shape == X.shape


# ---------------------------------------------------------------------------
# Metadata tables
# ---------------------------------------------------------------------------

class TestMetadata:

    def test_write_read_obs(self, tmp_path):
        A = make_sparse(m=100, n=50)
        obs = {
            "cell_type": ["A" if i % 2 == 0 else "B" for i in range(100)],
        }
        path = str(tmp_path / "meta.spz")
        stp.st_write(A, path, obs=obs, include_transpose=False)
        info = stp.st_info(path)
        assert info["has_obs"]
        obs_back = stp.st_read_obs(path)
        assert "cell_type" in obs_back
        assert len(obs_back["cell_type"]) == 100

    def test_no_metadata(self, tmp_path):
        A = make_sparse()
        path = str(tmp_path / "nometa.spz")
        stp.st_write(A, path, include_transpose=False)
        info = stp.st_info(path)
        assert not info["has_obs"]
        assert not info["has_var"]


# ---------------------------------------------------------------------------
# Compression check
# ---------------------------------------------------------------------------

class TestCompression:

    def test_file_smaller_than_raw(self, tmp_path):
        """SPZ file should be smaller than raw float32 CSC binary."""
        A = make_count_matrix(m=2000, n=200)
        path = str(tmp_path / "compress.spz")
        stp.st_write(A, path, include_transpose=False)
        spz_bytes = os.path.getsize(path)
        raw_bytes = A.nnz * 4  # float32
        assert spz_bytes < raw_bytes, (
            f"SPZ ({spz_bytes}) should be smaller than raw float32 ({raw_bytes})")
